lib/ua_update_messaging.py

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python3
+
+"""
+Update messaging text for use in MOTD and APT custom Ubuntu Advantage messages.
+
+Messaging files will be emitted to /var/lib/ubuntu-advantage/message-* which
+will be sourced by apt-hook/hook.cc and various /etc/update-motd.d/ hooks to
+present updated text about Ubuntu Advantage service and token state.
+"""
+
+import distro_info  # type: ignore
+import enum
+import logging
+import os
+
+try:
+    from typing import Optional  # noqa
+except ImportError:
+    # typing isn't available on trusty, so ignore its absence
+    pass
+
+from uaclient.cli import setup_logging
+from uaclient import config
+from uaclient import defaults
+from uaclient import status
+from uaclient import util
+
+
+# Type of message file used for external messaging (APT and MOTD)
+@enum.unique
+class ExternalMessage(enum.Enum):
+    CONTRACT_EXPIRY = "contract-expiry-status.tmpl"
+    ESM_ANNOUNCE = "esm-announce"
+
+
+def format_contract_expiry_message(cfg: config.UAConfig,) -> "Optional[str]":
+    """Return warning message about contract expiry used by MOTD and APT.
+
+    Compare UAConfig.contract_remainining_days versus pending expiry period,
+    expiry grace period and absolute expiry dates. Construct appropriate
+    message to emit in APT and MOTD based on expired contract state.
+
+    :param cfg: UAConfig instance for this environment.
+    """
+    if not cfg.is_attached:
+        return None
+    grace_period = -defaults.CONTRACT_EXPIRY_GRACE_PERIOD_DAYS
+    pending_expiry = defaults.CONTRACT_EXPIRY_PENDING_DAYS
+    remaining_days = cfg.contract_remaining_days
+    if 0 <= remaining_days <= pending_expiry:
+        return status.MESSAGE_CONTRACT_NEAR_EXPIRED_TMPL.format(
+            remaining_days=remaining_days, url=defaults.BASE_UA_URL
+        )
+    elif grace_period <= remaining_days < 0:
```

Comment:
It seems `grace_period`  is positive here. Shouldn't it be a negative number ?

Reply:
Sorry, probably confusing how I setup the local grace_period variable above as a negative. Probably easiest to make that postive/negative conversion locally here on the line instead of [up here ](https://github.com/canonical/ubuntu-advantage-client/pull/1542/files#diff-02bd06ac4b2af43bf296d5379d9a0d3c25cf70e5f417ea5d34fc9b98ca6683d6R47)

Reply:
I'll change that to inflect the value of local grace_period  here

Reply:
:facepalm: my mistake, just saw the negative sign there